File: src/exercises/chapter1/lesson3.rb
# frozen_string_literal: true

# S(n) = 1 + ½ + 1/3 + … + 1/n
# @param {Integer} n - The upper limit of the range
# @return {Float} - The sum of the harmonic series up to n

def solution(n)
  return 0 if n.zero?

  result = 0
  (1..n).each do |i|
    result += 1 / i.to_f
  end
  result
end

File: src/exercises/chapter1/Lesson3.ts
/**
 * Lesson 3: Calculate S(n) = 1 + ½ + 1/3 + … + 1/n
 */
export default class Lesson3 {

  /**
   * Calculates the sum of the harmonic series from 1 to n.
   *
   * @param {number} n - The upper limit of the harmonic series.
   * @return {number} The calculated sum S(n) = 1 + ½ + 1/3 + … + 1/n.
   */
  solution(n: number): number {
    // Initialize the result variable to store the sum.
    let result = 0;

    // Iterate through the numbers from 1 to n and add the inverse of each number to the result.
    for(let i = 1; i <= n; i++){
      result += 1 / i;
    }

    // Return the final calculated result.
    return result;
  }
}
